File: maps/src/main/java/it/localhostsoftware/maps/huawei/model/HuaweiMarkerOptions.java
package it.localhostsoftware.maps.huawei.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import it.localhostsoftware.maps.model.BitmapDescriptor;
import it.localhostsoftware.maps.model.LatLng;
import it.localhostsoftware.maps.model.MarkerOptions;

public class HuaweiMarkerOptions extends MarkerOptions<com.huawei.hms.maps.model.MarkerOptions> {
    public HuaweiMarkerOptions(com.huawei.hms.maps.model.MarkerOptions markerOptions) {
        super(markerOptions);
    }

    @Override
    public MarkerOptions<?> position(@NonNull LatLng<?> latLng) {
        getMarkerOptions().position((com.huawei.hms.maps.model.LatLng) latLng.getLatLng());
        return this;
    }

    @Override
    public MarkerOptions<?> zIndex(float v) {
        getMarkerOptions().zIndex(v);
        return this;
    }

    @Override
    public MarkerOptions<?> icon(@Nullable BitmapDescriptor<?> var1) {
        getMarkerOptions().icon(var1 == null ? null : (com.huawei.hms.maps.model.BitmapDescriptor) var1.getBitmapDescriptor());
        return this;
    }

    @Override
    public MarkerOptions<?> anchor(float v, float v1) {
        getMarkerOptions().anchorMarker(v, v1);
        return this;
    }

    @Override
    public MarkerOptions<?> infoWindowAnchor(float v, float v1) {
        getMarkerOptions().infoWindowAnchor(v, v1);
        return this;
    }

    @Override
    public MarkerOptions<?> title(@Nullable String s) {
        getMarkerOptions().title(s);
        return this;
    }

    @Override
    public MarkerOptions<?> snippet(@Nullable String s) {
        getMarkerOptions().snippet(s);
        return this;
    }

    @Override
    public MarkerOptions<?> draggable(boolean b) {
        getMarkerOptions().draggable(b);
        return this;
    }

    @Override
    public MarkerOptions<?> visible(boolean b) {
        getMarkerOptions().visible(b);
        return this;
    }

    @Override
    public MarkerOptions<?> flat(boolean b) {
        getMarkerOptions().flat(b);
        return this;
    }

    @Override
    public MarkerOptions<?> rotation(float v) {
        getMarkerOptions().rotation(v);
        return this;
    }

    @Override
    public MarkerOptions<?> alpha(float v) {
        getMarkerOptions().alpha(v);
        return this;
    }

    @Override
    public LatLng<?> getPosition() {
        return new HuaweiLatLng(getMarkerOptions().getPosition());
    }

    @Override
    public String getTitle() {
        return getMarkerOptions().getTitle();
    }

    @Override
    public String getSnippet() {
        return getMarkerOptions().getSnippet();
    }

    @Override
    public BitmapDescriptor<?> getIcon() {
        return new BitmapDescriptor<>(getMarkerOptions().getIcon());
    }

    @Override
    public float getAnchorU() {
        return getMarkerOptions().getMarkerAnchorU();
    }

    @Override
    public float getAnchorV() {
        return getMarkerOptions().getMarkerAnchorV();
    }

    @Override
    public boolean isDraggable() {
        return getMarkerOptions().isDraggable();
    }

    @Override
    public boolean isVisible() {
        return getMarkerOptions().isVisible();
    }

    @Override
    public boolean isFlat() {
        return getMarkerOptions().isFlat();
    }

    @Override
    public float getRotation() {
        return getMarkerOptions().getRotation();
    }

    @Override
    public float getInfoWindowAnchorU() {
        return getMarkerOptions().getInfoWindowAnchorU();
    }

    @Override
    public float getInfoWindowAnchorV() {
        return getMarkerOptions().getInfoWindowAnchorV();
    }

    @Override
    public float getAlpha() {
        return getMarkerOptions().getAlpha();
    }

    @Override
    public float getZIndex() {
        return getMarkerOptions().getZIndex();
    }
}

File: maps/src/main/java/it/localhostsoftware/maps/huawei/model/HuaweiMarkerOptions.kt
package it.localhostsoftware.maps.huawei.model

import com.huawei.hms.maps.model.MarkerOptions

class HuaweiMarkerOptions(markerOptions: MarkerOptions) : it.localhostsoftware.maps.model.MarkerOptions<MarkerOptions, HuaweiBitmapDescriptor, HuaweiLatLng>(markerOptions) {
    override var position: HuaweiLatLng
        get() = HuaweiLatLng(mo.position)
        set(value) {
            mo.position(value.ll)
        }
    override var title: String?
        get() = mo.title
        set(value) {
            mo.title(value)
        }
    override var snippet: String?
        get() = mo.snippet
        set(value) {
            mo.snippet(value)
        }
    override var icon: HuaweiBitmapDescriptor?
        get() = mo.icon?.let { HuaweiBitmapDescriptor(it) }
        set(value) {
            mo.icon(value?.bitmapDescriptor)
        }
    override var isDraggable: Boolean
        get() = mo.isDraggable
        set(value) {
            mo.draggable(value)
        }
    override var isVisible: Boolean
        get() = mo.isVisible
        set(value) {
            mo.visible(value)
        }
    override var isFlat: Boolean
        get() = mo.isFlat
        set(value) {
            mo.flat(value)
        }
    override var rotation: Float
        get() = mo.rotation
        set(value) {
            mo.rotation(value)
        }
    override var alpha: Float
        get() = mo.alpha
        set(value) {
            mo.alpha(value)
        }
    override var zIndex: Float
        get() = mo.zIndex
        set(value) {
            mo.zIndex(value)
        }
    override var anchor: Pair<Float, Float>
        get() = Pair(mo.markerAnchorU, mo.markerAnchorV)
        set(value) {
            mo.anchorMarker(value.first, value.second)
        }
    override var infoWindowAnchor: Pair<Float, Float>
        get() = Pair(mo.infoWindowAnchorU, mo.infoWindowAnchorV)
        set(value) {
            mo.infoWindowAnchor(value.first, value.second)
        }
}